frontend/src/lib/modals/sns/SwapModal/ParticipateSwapModal.svelte

```
@@ -106,6 +108,32 @@
       stopBusy("project-participate");
     }
   };
+
+  // Used for form inline validation
+  let validateAmount: (amount: number | undefined) => string | undefined;
+  $: validateAmount = (amount: number | undefined) => {
+    if (
+      swapCommitment !== undefined &&
+      swapCommitment !== null &&
+      amount !== undefined
+    ) {
+      const { valid, labelKey, substitutions } = validParticipation({
+        project: {
+          rootCanisterId: summary.rootCanisterId,
+          summary,
+          swapCommitment,
+        },
+        amount: TokenAmount.fromNumber({ amount, token: ICPToken }),
+      });
+      // `validParticipation` does not return `valid` as `false` without a labelKey.
+      // But we need to check because of type safety.
+      return valid || labelKey === undefined
```

Not sure about swallowing an error in case of valid=false and no labelKey. Probably showing some general error would improve this. Not a strong opinion.

---

That's a good point.

Also a variant maybe?

```
type ValidationResult =
  { Ok: null } |
  { Err: { labelKey: string, substitutions: ... } };
```